apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -0,0 +1,195 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Linq;
+using Google.Api.Gax;
+
+// ReSharper disable UnusedParameter.Local
+
+namespace Google.Cloud.Spanner
+{
+    /// <summary>
+    /// </summary>
+    public sealed class SpannerCommandTextBuilder
+    {
+        private static readonly string s_insertCommand = "INSERT";
+        private static readonly string s_insertUpdateCommand = "INSERTUPDATE";
+        private static readonly string s_updateCommand = "UPDATE";
+        private static readonly string s_deleteCommand = "DELETE";
+        private static readonly string s_selectCommand = "SELECT";
+
+        private string _targetTable;
+
+        /// <summary>
+        /// </summary>
+        /// <param name="commandText"></param>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="InvalidOperationException"></exception>
+        public SpannerCommandTextBuilder(string commandText)
+        {
+            GaxPreconditions.CheckNotNullOrEmpty(commandText, nameof(commandText));
+            var commandSections = commandText.Split(' ');
+            if (commandSections.Length < 2)
+                throw new InvalidOperationException($"{commandText} is not a recognized Spanner command.");
+            var newBuilder = new SpannerCommandTextBuilder();
+            if (!TryParseCommand(this, s_deleteCommand, SpannerCommandType.Delete, commandSections)
+                && !TryParseCommand(this, s_updateCommand, SpannerCommandType.Update, commandSections)
+                && !TryParseCommand(this, s_insertCommand, SpannerCommandType.Insert, commandSections)
+                && !TryParseCommand(this, s_insertUpdateCommand, SpannerCommandType.InsertOrUpdate, commandSections))
+            {
+                if (!commandSections[0].ToUpper().StartsWith(s_selectCommand))
+                    throw new InvalidOperationException($"{commandText} is not a recognized Spanner command.");
+                newBuilder.CommandText = commandText;
+                newBuilder.SpannerCommandType = SpannerCommandType.Select;
+            }
+        }
+
+        /// <summary>
+        /// </summary>
+        private SpannerCommandTextBuilder()
+        {
+        }
+
+        /// <summary>
+        /// </summary>
+        public string CommandText { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        public SpannerCommandType SpannerCommandType { get; private set; }
```

Comment:
Could this be made fully read-only by adding a constructor with suitable parameters? (Basically, whenever a property *shouldn't* change after initialization, I prefer to make that obvious.)

The same may be true for CommandText, but I'd originally assumed not...